ydb/library/yql/dq/opt/dq_opt_dphyp_solver.h

```diff
@@ -0,0 +1,473 @@
+};
+
+/*
+ * Count the number of items in the DP table of DPHyp
+ */
+template <typename TNodeSet> ui32 TDPHypSolver<TNodeSet>::CountCC(ui32 budget) {
+    TNodeSet allNodes;
+    allNodes.set();
+    ui32 cost = 0;
+
+    for (int i = NNodes_ - 1; i >= 0; --i) {
+        ++cost;
+        if (cost > budget) {
+            return cost;
+        }
+        TNodeSet s;
+        s[i] = 1;
+        TNodeSet x = MakeB(allNodes,i);
+        cost = CountCCRec(s, x, cost, budget);
+    }
+
+    return cost;
+}
+
+/**
+ * Recursively count the nuber of items in the DP table of DPccp
+*/
+template <typename TNodeSet> ui32 TDPHypSolver<TNodeSet>::CountCCRec(const TNodeSet& s, const TNodeSet& x, ui32 cost, ui32 budget) {
+    TNodeSet neighs = Neighs(s, x);
+
+    if (neighs == TNodeSet{}) {
+        return cost;
+    }
+
+    TNodeSet prev;
+    TNodeSet next;
+
+    while(true) {
+        next = NextBitset(prev, neighs);
+        cost += 1;
+        if (cost > budget) {
+            return cost;
+        }
+        cost = CountCCRec(s | next, x | neighs, cost, budget);
+        if (next == neighs) {
+            break;
+        }
+        prev = next;
+    }
+
+    return cost;
+}
+
+template<typename TNodeSet> TNodeSet TDPHypSolver<TNodeSet>::Neighs(TNodeSet s, TNodeSet x) {
+    TNodeSet neighs{};
+
+    auto& nodes = Graph_.GetNodes();
+
+    TSetBitsIt<TNodeSet> setBitsIt(s);
+    while (setBitsIt.HasNext()) {
+        size_t nodeId = setBitsIt.Next();
+        
+        neighs |= nodes[nodeId].SimpleNeighborhood;
+
+        for (const auto& edgeId: nodes[nodeId].ComplexEdgesId) {
+            auto& edge = Graph_.GetEdge(edgeId);
+            if (
+                IsSubset(edge.Left, s) &&
+                !AreOverlaps(edge.Right, x) &&
+                !AreOverlaps(edge.Right, s) && 
+                !AreOverlaps(edge.Right, neighs)
+            ) {
+                neighs[GetLowestSetBit(edge.Right)] = 1;
+            }
+        }
+    }
+
+    neighs &= ~x;
+    return neighs;
+}
+
+template<>
+inline std::bitset<64> TDPHypSolver<std::bitset<64>>::NextBitset(const std::bitset<64>& prev, const std::bitset<64>& final) {
+    return std::bitset<64>((prev | ~final).to_ulong() + 1) & final;
+}
+
+template<typename TNodeSet> TNodeSet TDPHypSolver<TNodeSet>::NextBitset(const TNodeSet& prev, const TNodeSet& final) {
+    if (prev == final) {
+        return final;
+    }
+
+    TNodeSet res = prev;
+
+    bool carry = true;
+    for (size_t i = 0; i < NNodes_; i++)
+    {
+        if (!carry) {
+            break;
+        }
+
+        if (!final[i]) {
+            continue;
+        }
+
+        if (res[i] == 1 && carry) {
+            res[i] = 0;
+        } else if (res[i] == 0 && carry) {
+            res[i] = 1;
+            carry = false;
+        }
+    }
+
+    return res;
+}
+
+template<typename TNodeSet> std::shared_ptr<TJoinOptimizerNodeInternal> TDPHypSolver<TNodeSet>::Solve() {
```

Comment:
Need a comment for this method. The name 'solve' is from a paper and is a poor description of what happens here